File: agents/imac/JarvisAgent/Sources/JarvisAgent/Tools/QuitApplicationTool.swift
// REQUIRES REAL macOS VALIDATION — NSRunningApplication cannot be
// exercised outside a real macOS runtime. Never compiled or run in this
// environment.

import AppKit

/// Quits a named, already-running application by asking it to terminate
/// via the OS's own public API — the reverse of OpenApplicationTool.
/// Deliberately NOT "kill any process by PID/name": only an application
/// NSWorkspace itself already knows is running gets asked to quit, the
/// same low-privilege spirit as the rest of this registry (see
/// ToolRegistry.swift) — there is no code path here that can terminate
/// anything other than a regular, user-visible app.
enum QuitApplicationTool {
    static func make() -> AgentTool {
        AgentTool(name: "quit_application") { input in
            guard let applicationName = input["applicationName"]?.value as? String,
                  !applicationName.trimmingCharacters(in: .whitespaces).isEmpty
            else {
                return ToolResultPayload(success: false, data: nil, error: "applicationName is required")
            }

            let runningApps = NSWorkspace.shared.runningApplications.filter { app in
                app.localizedName?.caseInsensitiveCompare(applicationName) == .orderedSame
            }

            guard !runningApps.isEmpty else {
                return ToolResultPayload(success: false, data: nil, error: "Application not running: \(applicationName)")
            }

            for app in runningApps {
                app.terminate()
            }

            let result: [String: AnyCodable] = ["application": AnyCodable(applicationName)]
            return ToolResultPayload(success: true, data: AnyCodable(result), error: nil)
        }
    }
}

File: src/tools/system/QuitApplicationTool.ts
import { PermissionLevel } from "@/types/permissions";
import type { DeviceTool } from "@/types/tools";

/**
 * Quits a named, already-running application on the target device — the
 * reverse of OPEN_APPLICATION. Runs only on a device agent; Core has no
 * local implementation. The Agent's own implementation only ever asks an
 * application NSWorkspace already knows is running to terminate via the
 * OS's public API, never an arbitrary process kill by PID/name.
 * SAFE_ACTION for the same reason as OPEN_APPLICATION: a real, visible
 * action, but reversible (the app can just be reopened) and low-impact.
 */
export const quitApplicationTool: DeviceTool = {
  id: "QUIT_APPLICATION",
  name: "quit_application",
  description:
    'Quits a named, currently running application on the target device (e.g. "Slack", "Spotify"). ' +
    "Use the application's actual display name. Fails if the application isn't currently running.",
  inputSchema: {
    type: "object",
    properties: {
      applicationName: {
        type: "string",
        description: 'The application\'s name, e.g. "Slack" or "Spotify".',
      },
      deviceId: {
        type: "string",
        description: "Device to quit the application on. Defaults to the primary device if omitted.",
      },
    },
    required: ["applicationName"],
  },
  requiredPermission: PermissionLevel.SAFE_ACTION,
  target: "device",
};
